### PR TITLE
Support ONNX Export for HF_TEXT

### PR DESCRIPTION
**Pull request by FANGAreNotGnu**
- Code changes to support ONNX export for HF_TEXT
- Write and example to export an ONNX model and also a sample test
- Add raw feature output (before pooling) for hf_text
- Add raw feature distillation

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: multimodal/src/autogluon/multimodal/utils/onnx.py
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/onnx.py:64]: `hf_text_text_token_ids_column_sentence1 ` is not in `valid_input`. Will it be used in the onnx export?

File: multimodal/src/autogluon/multimodal/utils/onnx.py
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/onnx.py:46]: Why do we need to add the `hf_text` as the prefix in each key? It looks kind of redundant.

File: multimodal/src/autogluon/multimodal/data/process_text.py
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/data/process_text.py:247]: We'd better not remove the two \newlines. Without them, the docstring format changes.

File: examples/automm/production/hf_to_onnx_example.py
Comment (zhiqiangdon) [inline examples/automm/production/hf_to_onnx_example.py:32]: Do we this function? MultimodalPredictor already has the API.

File: multimodal/src/autogluon/multimodal/predictor.py
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/predictor.py:2012]: Do we need the `batch` argument? Having the `data` argument seems enough. Having both could cause confusions and complex the logic.

File: multimodal/src/autogluon/multimodal/predictor.py
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/predictor.py:2048]: How about removing these two lines and directly pass `self._model.eval()` to `torch.onnx.export()`?

File: multimodal/src/autogluon/multimodal/predictor.py
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/predictor.py:2060]: If we remove the `batch` argument, we could remove this part to simplify the logic.

Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/predictor.py:2060]: If users provide data, we use their data. Otherwise, use our default batch.

File: multimodal/src/autogluon/multimodal/predictor.py
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/predictor.py:2129]: This part is the same as that in `_realtime_predict()`. Could we wrap it into a shared function?

File: multimodal/src/autogluon/multimodal/predictor.py
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/predictor.py:2137]: If `onnx_tracing ` and `to_numpy ` are both false, it would result in an empty return. Is `to_numpy` necessary?

File: multimodal/src/autogluon/multimodal/predictor.py
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/predictor.py:2136]: This transformation doesn't work for image models. We may need to convert the inputs based on the model types.

File: multimodal/src/autogluon/multimodal/predictor.py
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/predictor.py:2141]: When do we need `onnx_tracing` is False?

File: multimodal/src/autogluon/multimodal/configs/distiller/default.yaml
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/configs/distiller/default.yaml:14]: Consider putting these changes into another PR since they not related to onnx export.

File: multimodal/src/autogluon/multimodal/utils/onnx.py
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/utils/onnx.py:64]: It's used in tracing but is treated as constants onnx's graph. It only contributes to column feature but we extract only feature, so it won't affect ONNX's output.

Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/utils/onnx.py]: It's used in a previous version. Removed.

File: multimodal/src/autogluon/multimodal/utils/onnx.py
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/utils/onnx.py:46]: It's the default key names in hf input batch. I didn't change it while export onnx.

File: multimodal/src/autogluon/multimodal/data/process_text.py
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/data/process_text.py]: Done, thanks!

File: examples/automm/production/hf_to_onnx_example.py
Comment (FANGAreNotGnu) [reply examples/automm/production/hf_to_onnx_example.py]: Removed the whole file and created a new example.

File: multimodal/src/autogluon/multimodal/predictor.py
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/predictor.py]: Done, thanks!

File: multimodal/src/autogluon/multimodal/predictor.py
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/predictor.py]: batch was used for development. Just removed and we could add it back if we later find it useful

File: multimodal/src/autogluon/multimodal/predictor.py
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/predictor.py]: Done.

File: multimodal/src/autogluon/multimodal/predictor.py
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/predictor.py:2121]: I'll move it to a util function while working on CLIP

File: multimodal/src/autogluon/multimodal/predictor.py
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/predictor.py:2128]: This line is for when we want to evalute onnx/torchscript models but the batch size is not a dynamic dimension. But currently is not necessary.

File: multimodal/src/autogluon/multimodal/configs/distiller/default.yaml
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/configs/distiller/default.yaml]: Done.

File: multimodal/src/autogluon/multimodal/predictor.py
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/predictor.py:2122]: Added the missing condition. If onnx_tracing and to_numpy is false it will output raw tensor without type changes.

File: multimodal/src/autogluon/multimodal/predictor.py
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/predictor.py]: Removed. It was used for development

File: multimodal/src/autogluon/multimodal/utils/onnx.py
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/utils/onnx.py:46]: Got it!

File: multimodal/src/autogluon/multimodal/predictor.py
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/predictor.py:2073]: This function name seems not accurate enough. How about `get_onnx_batch` or `process_onnx_batch`?

File: multimodal/src/autogluon/multimodal/models/huggingface_text.py
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/huggingface_text.py:205]: Where do we use `RAW_FEATURES`? It seems unused in this PR.

File: multimodal/src/autogluon/multimodal/predictor.py
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/predictor.py:2020]: Do we need a unit test for the onnx export? Maybe some code and data from the `production/onnx_text.py`.

File: multimodal/src/autogluon/multimodal/predictor.py
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/predictor.py:2073]: It may be also used in torchscript in the future, so I'm trying not to add onnx here.

Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/predictor.py:2073]: `get_processed_batch ` sounds a little too general. I guess this function is only to convert model for deployment/production.

File: examples/automm/pipeline/feature_extraction_example.py
Comment (sxjscience) [inline examples/automm/pipeline/feature_extraction_example.py:75]: Should we just use `sentence-transformers_msmarco-MiniLM-L-12-v3.onnx` ?

Comment (FANGAreNotGnu) [reply examples/automm/pipeline/feature_extraction_example.py]: Here I'm trying to show two ways of feature extraction: MultiModalPredictor v.s. onnx. checkpoint_name is for MultiModalPredictor. I updated the code to infer a default onnx_path using checkpoint_name.
Also I'd like to update that there is a large inference time gap between onnx and MultiModalPredictor, for this STSBenchmark task the time difference (model=minilm-l-12) is:
Autogluon load time: 3.8375
Autogluon eval time: 16.1105
ONNX load time: 1.2615
ONNX eval time: 2.2466

File: multimodal/src/autogluon/multimodal/predictor.py
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/predictor.py:2020]: Done.

File: multimodal/src/autogluon/multimodal/models/huggingface_text.py
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/models/huggingface_text.py]: Removed.

**Review by zhiqiangdon — APPROVED** [review]
LGTM! Thanks!  Need to resolve the conflicts before merge.

File: examples/automm/pipeline/feature_extraction_example.py
Comment (sxjscience) [inline examples/automm/pipeline/feature_extraction_example.py:18]: Avoid using `eval` as the name for function, since eval is a built-in function in python.

Change it to `evaluate(..)`.

Comment (sxjscience) [reply examples/automm/pipeline/feature_extraction_example.py:18]: @FANGAreNotGnu need to change here also.

**Review by sxjscience — APPROVED** [review]
LGTM! A small comment regarding eval —> evaluate.

**Comment by sxjscience** [comment]
I just noticed a conflict file. We can merge this in once the “eval” is replaced with “evaluate” and the  conflicting is resolved.

**Review by bryanyzhu — APPROVED** [review]
Looks good, thank you.

**Pull request merged**
